kuzzle-esp32/01-rgb-led/rgb-led.md

```diff
@@ -0,0 +1,333 @@
+
+        cJSON* on = cJSON_GetObjectItem(jstate, "on");
+        if (on != NULL)
+            _light_state.on = on->valueint;
+
+        ESP_LOGD(TAG,
+                 "New light state: r= 0x%02x, g= 0x%02x, b= 0x%02x, on = %s",
+                 _light_state.r,
+                 _light_state.g,
+                 _light_state.b,
+                 _light_state.on ? "true" : "false");
+        _update_light();
+    } 
+}
+
+static void _update_light()
+{
+    uint32_t r_duty = _light_state.on ? (LEDC_MAX_PWM * _light_state.r) / 0xFF : 0;
+    uint32_t g_duty = _light_state.on ? (LEDC_MAX_PWM * _light_state.g) / 0xFF : 0;
+    uint32_t b_duty = _light_state.on ? (LEDC_MAX_PWM * _light_state.b) / 0xFF : 0;
+
+    ledc_set_fade_with_time(LEDC_HIGH_SPEED_MODE, RED_PWM_CHANNEL, r_duty, LEDC_TRANSITION_TIME);
+    ledc_set_fade_with_time(LEDC_HIGH_SPEED_MODE, GREEN_PWM_CHANNEL, g_duty, LEDC_TRANSITION_TIME);
+    ledc_set_fade_with_time(LEDC_HIGH_SPEED_MODE, BLUE_PWM_CHANNEL, b_duty, LEDC_TRANSITION_TIME);
+
+    ledc_fade_start(LEDC_HIGH_SPEED_MODE, RED_PWM_CHANNEL, LEDC_FADE_NO_WAIT);
+    ledc_fade_start(LEDC_HIGH_SPEED_MODE, GREEN_PWM_CHANNEL, LEDC_FADE_NO_WAIT);
+    ledc_fade_start(LEDC_HIGH_SPEED_MODE, BLUE_PWM_CHANNEL, LEDC_FADE_NO_WAIT);
+
+    _publish_light_state(); // Publish new complete state
+}
+```
+
+## Get the Code
+
+You can get the whole source code for the rgb light from my github repository: <https://github.com/etrousset/instructable.git>. It's located on the 'rbg-light' branch. You can get it using the following git command:
+
+``` console
+$ git clone -b rgb-light https://github.com/etrousset/instructable.git
+```
+
+## Visualize Device State
+
+![visualise device state](img/admin-console-device-state.png)
+
+For this last step we will use Kuzzle Console to visualize and change the state of the RGB light.
+
+When booting, the device will publish its current state. You should be able to visualize it in Kuzzle Console as shown in the picture.
+
+To do so, open Kuzzle Console and connect it to your Kuzzle server instance. Then open the **iot** index, and then the **device-state** collection.
+
+If your device booted up, you should see a document representing the state of your RBG light. The last step is to control the RBG light through Kuzzle server. The easiest way to do this for now is using Kuzzle Console.
+
+## Get in Control...
+
+In this last step, we'll change the color of the RGB light by creating a new document in device-state collection.
+
+We will mark this document avec being a 'partial_state', meaning the device will have to merge it with its current state in order to get its new complete state. This allow changing only one parameter of the state with out having to publish the whole state. Also, the device is responsible for accepting or not the state change.
+
+In the case where it will update its state (turning ON or OFF, changing color) it will then publish its new complete state.
```

Comment:
Following my last 2 comments, a proposal: `In return, the device will send to Kuzzle its complete state, updated.`